### PR TITLE
[mlir][Vector] Add a rewrite pattern for better low-precision bitcast…

### PR DESCRIPTION
**Pull request by nicolasvasilache**
…(trunci) expansion

This revision adds a rewrite for sequences of vector `bitcast(trunci)` to use a more efficient sequence of vector operations comprising `shuffle` and `bitwise` ops.

Such patterns appear naturally when writing quantization / dequantization functionality with the vector dialect.

The rewrite performs a simple enumeration of each of the bits in the result vector and determines its provenance in the pre-trunci vector. The enumeration is used to generate the proper sequence of `shuffle`, `andi`, `ori` followed by an optional final `trunci`/`extui`.

The rewrite currently only applies to 1-D non-scalable vectors and bails out if the final vector element type is not a multiple of 8. This is a failsafe heuristic determined empirically: if the resulting type is not an even number of bytes, further complexities arise that are not improved by this pattern: the heavy lifting still needs to be done by LLVM.

### CODE REVIEW

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 1]
```diff
@@ -155,6 +164,221 @@ struct ConvertVectorTransferRead final
...
+
+    // BitCastBitsEnumerator encodes for each element of the target vector the
+    // provenance of the bits in the source vector. We can "transpose" this
+    // information to build a sequence of shuffles and bitwise ops that will
+    // produce the desired result.
+    // The algorithm proceeds as follows:
+    //   1. there are as many shuffles as max entries in BitCastBitsEnumerator
+    //   2. for each shuffle:
+    //     a. collect the source vectors that participate in this shuffle. One
+    //     source vector per target element of the shuffle. If overflow, take 0.
```
Comment (qcolombet) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:304]: What does it mean "if overflow"?

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 2]
```diff
@@ -155,6 +164,221 @@ struct ConvertVectorTransferRead final
...
+    //     b. the bitrange in the source vector as a mask. If overflow, take 0.
+    //     c. the number of bits to shift right to align the source bitrange at
+    //     position 0. This is exactly the low end of the bitrange.
+    //     d. number of bits to shift left to align to the desired position in
+    //     the result element vector.
```
Comment (qcolombet) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:309]: The comment is difficult to read in that `a.` is an action and `b.`-`d.` are a list of what we need.

**Review by qcolombet — COMMENTED** [review]
Partial review.
I still need to look at the meat of the transform, i.e., the actual rewrite.

So far, things look good, I suggested a few places where we could add comments.

**Comment by nicolasvasilache** [comment]
> So far, things look good, I suggested a few places where we could add comments.

Thanks! Added a bunch more comments, after the next PR many will likely end up living on the `BitCastBitsEnumerator` as I expect I can just use one impl with minor variations for other cases too.

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 3]
```diff
@@ -155,6 +164,280 @@ struct ConvertVectorTransferRead final
...
+//     source element contributing bits for the current vector.shuffle, take 0.
+//   3. shift right by the proper amount to align the source bitrange at
+//     position 0. This is exactly the low end of the bitrange. For instance,
+//     the first element of row 2 is `{ 1: b@[3..5) lshl: 0}` and one needs to
+//     shift right by 3 to get the bits contributed by the source element #1
+//     into position 0.
+//   4. shift left by the proper amount to to align to the desired position in
+//     the result element vector.  For instance, the contribution of the second
+//     source element for the first row needs to be shifted by `5` to form the
+//     first i8 result element.
```
Comment (qcolombet) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:337]: I think a summary of what the whole explanation is describing would be good. At first I didn't get why we needed 3 shuffles.
E.g.,
```
// In other words, this algorithm populates the bits like so:
// ```
//     src bits 0 ... 
// 1st shuffle |xxxxx   |xx      |...
// 2nd shuffle |     xxx|  xxxxx |...
// 3rd shuffle |        |       x|...
```

**Review by qcolombet — APPROVED** [review]
Food for thoughts: I wonder if this rewrite is at all useful if the input type of the `trunci` is as hostile for the LLVM backend (w.r.t. legalization and lowering) as the output type of `trunci`.

E.g., Does it make sense to do this rewrite for `trunci i13 to i5`?

**Comment by nicolasvasilache** [comment]
> Food for thoughts: I wonder if this rewrite is at all useful if the input type of the `trunci` is as hostile for the LLVM backend (w.r.t. legalization and lowering) as the output type of `trunci`.
> 
> E.g., Does it make sense to do this rewrite for `trunci i13 to i5`?

Indeed, this is what the following captures:
```
    // TODO: consider relaxing this restriction in the future if we find ways
    // to really work with subbyte elements across the MLIR/LLVM boundary.
    int64_t resultBitwidth = targetVectorType.getElementTypeBitWidth();
    if (resultBitwidth % 8 != 0)
      return rewriter.notifyMatchFailure(bitCastOp, "bitwidth is not k * 8");
```

**Pull request merged**